sw/airborne/firmwares/rotorcraft/stabilization/stabilization_attitude_quat_indi.c

```diff
@@ -0,0 +1,305 @@
+
+  stabilization_attitude_ref_enter();
+
+  FLOAT_RATES_ZERO(filtered_rate);
+  FLOAT_RATES_ZERO(filtered_rate_deriv);
+  FLOAT_RATES_ZERO(filtered_rate_2deriv);
+  FLOAT_RATES_ZERO(angular_accel_ref);
+  FLOAT_RATES_ZERO(indi_u);
+  FLOAT_RATES_ZERO(indi_du);
+  FLOAT_RATES_ZERO(u_act_dyn);
+  FLOAT_RATES_ZERO(u_in);
+  FLOAT_RATES_ZERO(udot);
+  FLOAT_RATES_ZERO(udotdot);
+
+}
+
+void stabilization_attitude_set_failsafe_setpoint(void) {
+  /* set failsafe to zero roll/pitch and current heading */
+  int32_t heading2 = stabilization_attitude_get_heading_i() / 2;
+  PPRZ_ITRIG_COS(stab_att_sp_quat.qi, heading2);
+  stab_att_sp_quat.qx = 0;
+  stab_att_sp_quat.qy = 0;
+  PPRZ_ITRIG_SIN(stab_att_sp_quat.qz, heading2);
+}
+
+void stabilization_attitude_set_rpy_setpoint_i(struct Int32Eulers *rpy) {
+  // stab_att_sp_euler.psi still used in ref..
+  memcpy(&stab_att_sp_euler, rpy, sizeof(struct Int32Eulers));
+
+  quat_from_rpy_cmd_i(&stab_att_sp_quat, &stab_att_sp_euler);
+}
+
+void stabilization_attitude_set_earth_cmd_i(struct Int32Vect2 *cmd, int32_t heading) {
+  // stab_att_sp_euler.psi still used in ref..
+  stab_att_sp_euler.psi = heading;
+
+  // compute sp_euler phi/theta for debugging/telemetry
+  /* Rotate horizontal commands to body frame by psi */
+  int32_t psi = stateGetNedToBodyEulers_i()->psi;
+  int32_t s_psi, c_psi;
+  PPRZ_ITRIG_SIN(s_psi, psi);
+  PPRZ_ITRIG_COS(c_psi, psi);
+  stab_att_sp_euler.phi = (-s_psi * cmd->x + c_psi * cmd->y) >> INT32_TRIG_FRAC;
+  stab_att_sp_euler.theta = -(c_psi * cmd->x + s_psi * cmd->y) >> INT32_TRIG_FRAC;
+
+  quat_from_earth_cmd_i(&stab_att_sp_quat, cmd, heading);
+}
+
+#define BOUND_CONTROLS(_v, _min, _max) {         \
+_v = _v < _min ? _min : _v > _max ? _max : _v;  \
+}
+
+static void attitude_run_indi(int32_t indi_commands[], struct Int32Quat *att_err)
+{
+  //Calculate required angular acceleration
+  struct FloatRates *body_rate = stateGetBodyRates_f();
+  angular_accel_ref.p = reference_acceleration.err_p * QUAT1_FLOAT_OF_BFP(att_err->qx)
+                      - reference_acceleration.rate_p * body_rate->p;
+  angular_accel_ref.q = reference_acceleration.err_q * QUAT1_FLOAT_OF_BFP(att_err->qy)
+                      - reference_acceleration.rate_q * body_rate->q;
+  angular_accel_ref.r = reference_acceleration.err_r * QUAT1_FLOAT_OF_BFP(att_err->qz)
+                      - reference_acceleration.rate_r * body_rate->r;
+
+  //Incremented in angular acceleration requires increment in control input
+  indi_du.p = 1.0/g1.p * (angular_accel_ref.p - filtered_rate_deriv.p);
+  indi_du.q = 1.0/g1.q * (angular_accel_ref.q - filtered_rate_deriv.q);
+  indi_du.r = 1.0/(g1.r+g2) * (angular_accel_ref.r - filtered_rate_deriv.r + g2*indi_du.r);
+
+  //add the increment to the total control input
+  u_in.p = indi_u.p + indi_du.p;
+  u_in.q = indi_u.q + indi_du.q;
+  u_in.r = indi_u.r + indi_du.r;
+
+  //bound the total control input
+  BOUND_CONTROLS(u_in.p, -4500, 4500);
```

Comment:
I couldn't find the definition of BOUND anywhere, I think that is why I chose to just write one myself... just to be sure